src/main/java/com/google/jenkins/plugins/k8sengine/KubernetesEnginePublisher.java

```diff
@@ -315,15 +338,57 @@ public FormValidation doCheckZone(
       return FormValidation.ok();
     }
 
-    public FormValidation doCheckProjectId(@QueryParameter String value) {
-      if (Strings.isNullOrEmpty(value)) {
+    public ListBoxModel doFillProjectIdItems(
+        @AncestorInPath Jenkins context,
+        @QueryParameter("credentialsId") final String credentialsId) {
+      ListBoxModel items = new ListBoxModel();
+      items.add("");
+      if (Strings.isNullOrEmpty(credentialsId)) {
+        return items;
+      }
+
+      ClientFactory clientFactory;
+      try {
+        clientFactory = this.getClientFactory(context, credentialsId);
+      } catch (AbortException ae) {
+        items.clear();
+        items.add(ae.getMessage(), "");
+        return items;
+      }
+
+      String defaultProjectId = clientFactory.getDefaultProjectId();
+      try {
+        CloudResourceManagerClient client = clientFactory.cloudResourceManagerClient();
+        List<Project> projects = client.getAccountProjects();
+
+        if (!projects.isEmpty()) {
```

Comment:
Returning early is preferred to nested if structure for improved readability. Suggest reversing the condition and returning early.

Reply:
Done.